### ISSUE TITLE
Possible to maintain a global history as well as local histories?

### ISSUE THREAD

**Issue by tanc** [fish-shell/fish-shell#7110]
fish version: `3.1.2`
os: `Manjaro 5.6.15-1`
terminal: `Terminator 1.92`

I'd like to modify the up-or-search functionality in the following way and hoping someone might be able to guide me:

1. Each shell session maintains it's own local history
2. There is a also global history file that is updated whenever a command is run
3. Pressing up while at cursor position 0 does a standard `commandline -f history-search-backward` but of the current shell's local history
4. If the cursor is at a position greater than 0 then search the global history with `commandline -f history-search-backward`

My reasoning is this: When I have multiple shells sessions open (always) it makes sense to me to be able to rapidly cycle back through my last few commands with the up arrow as I often need to repeat a recently used command. On the other hand I might want to use a command I used previously, either in my current session or in a past session and also quite possible in a current session in another shell. For this I'd like to be able to search across a 'global' history of commands, but only if I start typing a part of that command. Having all of the open shell session's history merged confuses me when I hit the up arrow rapidly to cycle back through a few commands, but it is also really useful when I want to search for a recently used command from a different active shell session.

At the moment I have the following:

```
function up-or-search -d "Depending on cursor position and current mode, either search backward or move up one line"
    # If we are already in search mode, continue        
    if commandline --search-mode    
        commandline -f history-search-backward    
        return    
    end    
           
    # If we are navigating the pager, then up always navigates    
    if commandline --paging-mode    
        commandline -f up-line      
        return    
    end    
           
    # We are not already in search mode.    
    # If we are on the top line, start search mode,    
    # otherwise move up    
    set lineno (commandline -L)    
    set cursorpos (string length (commandline -cp))    
                                                                                            
    switch $lineno                                                                          
        case 1                                                                              
            if [ $cursorpos = 0 ]                                                           
                commandline -f history-search-backward                                      
            else                                                                            
                history merge                                                               
                commandline -f history-search-backward                                      
            end                                                                             
                                                                                            
        case '*'                                                                            
            commandline -f up-line                                                          
    end                                                                                     
end
```

This is based off a comment in #825 but modified so that the `history merge` only happens if the cursor is not at the beginning of the line. While this works for now it would be interesting to try having separate histories as well as a global one to avoid the 'pollution' of the current session history.

I have seen that you can set the `fish_history` variable which would write the history to a different file. I guess it might be possible to use the default history as a global history (and keep it updated with constant `history merge` commands) and keep a local history by defining `fish_history` per session. Then switch between the default history when searching and a local history when using the up arrow at cursor position 0.

**Comment by tanc** [comment]
I've been experimenting a bit. I've been able to set a local history by adding `set -x fish_history $fish_pid` to my fish config so that when a new shell session is started the history is written to a file like `~/.local/share/fish/19449_history`

What I can't seem to do is get the `up-or-search` to use the default history temporarily:

```
function up-or-search -d "Depending on cursor position and current mode, either search backward or move up one line"
    set cursorpos (string length (commandline -cp))
                                                        
    # If we are already in search mode, continue
    if commandline --search-mode
        if [ $cursorpos = 0 ]
            commandline -f history-search-backward
            return                                                                                       
        else                                                                                             
            set fish_history "default"                                                                   
            commandline -f history-search-backward                                                       
            set fish_history $fish_pid                                                                   
            return                                                                       
        end                                                                                              
    end                                                                                  
                                                                                                         
    # If we are navigating the pager, then up always navigates                                           
    if commandline --paging-mode                                                                         
        commandline -f up-line                                                                           
        return                                                                                           
    end                                                                                                  
                                                                                                         
    # We are not already in search mode.                                                                 
    # If we are on the top line, start search mode,                                                      
    # otherwise move up                                                                                  
    set lineno (commandline -L)                                                                          
                                                                                                         
    switch $lineno                                                                                       
        case 1                                                                                           
            if [ $cursorpos = 0 ]                                                                        
                commandline -f history-search-backward                                                   
            else                                                                                         
                set fish_history "default"         
                history merge                                                                            
                commandline -f history-search-backward                                                   
            end                                                                                          
                                                                                                         
        case '*'                                                                                         
            commandline -f up-line                                                                       
    end                                                                                                  
    set fish_history $fish_pid                                                               
end
```

I get no output when the cursor is not position 0 and I'm searching something that I know is in the default history file.

My question at this point is: is it possible to use `history-search-backward` on a different history file in a fish function?

**Comment by tanc** [comment]
I think I've got it working now:

```
function up-or-search -d "Depending on cursor position and current mode, either search backward or move up one line"
    set cursorpos (string length (commandline -cp))    
    
    # If we are already in search mode, continue    
    if commandline --search-mode    
        if [ $cursorpos = 0 ]    
            set fish_history $fish_pid    
            commandline -f history-search-backward    
            return    
        else    
            set fish_history "default"    
            commandline -f history-search-backward    
            return    
        end    
    end        
           
    # If we are navigating the pager, then up always navigates    
    if commandline --paging-mode    
        commandline -f up-line      
        return    
    end    
           
    # We are not already in search mode.    
    # If we are on the top line, start search mode,    
    # otherwise move up    
    set lineno (commandline -L)    
        
    switch $lineno    
        case 1        
            if [ $cursorpos = 0 ]    
                set fish_history $fish_pid    
                commandline -f history-search-backward    
            else    
                set fish_history "default"    
                history merge    
                commandline -f history-search-backward    
            end    
                                               
        case '*'    
            commandline -f up-line    
    end    
           
end
```
It seems I didn't need to set the `fish_history` back to the value of the `$fish_pid` and in fact that was stopping it from working on the global history.

This now needs some testing and some way of cleaning up the `{$fish_pid}_history` files, maybe a cron task.

**Comment by tanc** [comment]
I'm getting close to admitting defeat on this. I was having trouble returning to the `$fish_pid` based history after `commandline -f history-search-backward`. If I have this:
```
set fish_history "default"    
history merge  
commandline -f history-search-backward
set fish_history $fish_pid
```
the default history is not searched. If I remove `set fish_history $fish_pid` then the default history is searched but after hitting return the `$fish_pid` based history is not returned to and future history writes happen on the default history.

I was able to get around that with a function listening to the `fish_postexec` event:
```
function on_fish_postexec --on-event fish_postexec
    set -x fish_history $fish_pid
end
```
The problem now is that the local `$fish_pid` based history is not updated when executing a command from the history search, it still writes to the default history. I've tried `history save` after setting the `fish_history` to the `$fish_pid` but that doesn't work. I've even tried manually writing to the local history file but fish appears to ignore that, it maybe retains the session history in memory?

Can anyone who understands how the history works shed some light on this? Is this possible or should I give up?

**Comment by zanchey** [comment]
There is no simple way to write a command to two history files at once, which I think is the core issue here. You could write it by hand to a local history and it should be reloaded when you switch, but the behaviour gets quite difficult to reason about. Autosuggestions will not behave predictably under your current code, I think.

**Comment by tanc** [comment]
I suspected as much. I've tried writing out the local history file by hand but that doesn't seem to get loaded when switching to it. I'm guessing there is a session history that is cached somewhere.

Well I think I'm going to close this and go back to what I had first which doesn't merge history if cursor is at position 0 which is part of the way there and lets me quickly scroll back through the last few commands in that session but still allows for merging of other sessions history.
